### PR TITLE
Simplify gather predicate generation.

### PR DESCRIPTION
**Pull request by naoyam**
This is a follow-up cleanup PR for #1307, which should be merged before this.

There should be no change in generated code. Just simplifying the logic to generate predicates for gather.

The main change is that when generating predicates for gather, we currently generate two predicates: one using the consumer index and another using the producer index. For example, suppose `x` be a consumer index, its producer index would look like: `x - left_pad + window_index`. Now, `left_pad` is just `int`, but `window_index` is a loop index variable, so it's not possible to statically determine which is larger. This means that we would need to have two predicates, one using the consumer index and another using the producer index. This is why `adjustStartAndStopOffsetsForGather` takes a vector of `kir::Val*`. 

In this PR, instead of generating two predicates for each of start and stop conditions, minimum or maximum of the producer and consumer offsets are used to generate just a single predicate. In addition, when statically analyzable, only either of the producer or consumer offsets is used.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/arith.cpp [hunk 1]
```diff
@@ -1344,6 +1348,10 @@ TensorView* gather(
     TORCH_CHECK(
         p.size() == 2,
         "Each entry of pad_width must have two non-negative integers.");
+    std::for_each(p.begin(), p.end(), [](const auto& p_left_or_right) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/arith.cpp:1336]: Should pad also not exceed (window_shape-1)/2?

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 2]
```diff
@@ -2810,20 +2793,30 @@ std::pair<std::vector<kir::Val*>, std::vector<kir::Val*>> getStartAndStopOffsets
...
+  stop_offset = ir_builder.subExpr(stop_offset, limits.second);
 
-  return {start_offsets, stop_offsets};
+  return {start_offset, stop_offset};
 }
 
-bool canOmitStartPredicate(kir::Val* start_offset) {
+// A partial value of a start offset is returned if determined to be
+// safe. Nullptr is returned if it can be omitted completely.
+kir::Val* simplifyStartPredicate(kir::Val* start_offset) {
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp]: Should this be `simplifyStartOffset`?

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/arith.cpp (hunk 1 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/arith.cpp:1336]: Yes, I'll add more comprehensive checks in an upcoming padding PR.

**Pull request merged**